### PR TITLE
Ml remove graph step

### PR DESCRIPTION
**Pull request by baykaner**

### CODE REVIEW

**Review by MissingNO57 — APPROVED** [review]
LGTM

File: libs/ml/tests/ml/layers/PRelu.cpp [hunk 1]
```diff
@@ -230,14 +230,16 @@ TYPED_TEST(PReluTest, saveparams_test)
   TypeParam loss = layer.Evaluate(error_output);
   layer.BackPropagate(error_output);
   layer.ApplyRegularisation();
-  layer.Step(DataType{0.1f});
```
Comment (evsmithx) [inline libs/ml/tests/ml/layers/PRelu.cpp:233]: With Step there was the option to set the learning rate - is there an equivalent with GetGradients + ApplyGradients?

**Review by evsmithx — DISMISSED** [review]

File: libs/ml/tests/ml/layers/PRelu.cpp [hunk 2]
```diff
@@ -230,14 +230,16 @@ TYPED_TEST(PReluTest, saveparams_test)
...
+  layer.ApplyGradients(grads);
 
   // train g2
   layer2.SetInput(label_name, labels);
   TypeParam loss2 = layer2.Evaluate(error_output);
   layer2.BackPropagate(error_output);
   layer2.ApplyRegularisation();
-  layer2.Step(DataType{0.1f});
+  auto grads2 = layer2.GetGradients();
+  layer2.ApplyGradients(grads2);
```
Comment (MissingNO57) [inline libs/ml/tests/ml/layers/PRelu.cpp:250]: I've just realised that you need to multiply gradient by -LEARNING_RATE otherwise you are maximising the error.

**Review by MissingNO57 — CHANGES_REQUESTED** [review]

File: libs/ml/tests/ml/layers/PRelu.cpp (hunk 2 above)
Comment (MissingNO57) [reply libs/ml/tests/ml/layers/PRelu.cpp:250]: Or use Optimiser which would do it for you.

Comment (baykaner) [reply libs/ml/tests/ml/layers/PRelu.cpp:250]: oh that's true. it actually makes no difference for this test; but it's good to demonstrate correct usage :-p I'll fix that

File: libs/ml/tests/ml/layers/PRelu.cpp (hunk 1 above)
Comment (baykaner) [reply libs/ml/tests/ml/layers/PRelu.cpp:233]: no. but there shouldn't be.
This is because Step was effectively doing SGD without pointing that out explicitly.

If we really want to do SGD without using the SGDOptimiser there's nothing that stops us doing what @MissingNO57 proposed (e.g. graph.ApplyGradients(-LEARNING_RATE * graph.GetGradients())), but then at least we're being explicit about what we're doing.

**Review by MissingNO57 — DISMISSED** [review]
LGTM

**Review by MissingNO57 — DISMISSED** [review]
LGTM

**Review by MissingNO57 — DISMISSED** [review]
LGTM?

**Review by MissingNO57 — APPROVED** [review]
LGTM!
